### ISSUE TITLE
NPE when applying CDK aromaticity model to an atomContainer having pseudoAtom with a double or triple bond

### ISSUE THREAD

**Issue by marois** [cdk/cdk#648]
I know sounds like a real edge case but many drawing tools expand abbreviation when producing CTAB format (e.g. v2000 or v3000 MDL formats). For example, drawing Acetonitrile with the nitrile using a nickname label CN will throw an NPE when checking aromaticity using the cdk legacy electron donation model. Note: applying the Daylight model does not NPE.

CDK version: 2.3

```
java.lang.NullPointerException
	at org.openscience.cdk.aromaticity.AtomTypeModel.contribution(AtomTypeModel.java:138)
	at org.openscience.cdk.aromaticity.Aromaticity.findBonds(Aromaticity.java:187)
	at org.openscience.cdk.aromaticity.Aromaticity.apply(Aromaticity.java:238)
```

```
    @Test
    public void aromaticity_should_not_throws_NPE_when_applying_CDK_ElectronDonation_model() throws CDKException {
        final String v3000Mol = "Acetonitrile\n" +
                "\n" +
                "\n" +
                "  0  0  0     0  0            999 V3000\n" +
                "M  V30 BEGIN CTAB\n" +
                "M  V30 COUNTS 3 2 1 0 0\n" +
                "M  V30 BEGIN ATOM\n" +
                "M  V30 1 C 64.4346 -22.1251 0 0\n" +
                "M  V30 2 C 65.9747 -22.1251 0 0\n" +
                "M  V30 3 N 67.5145 -22.1251 0 0\n" +
                "M  V30 END ATOM\n" +
                "M  V30 BEGIN BOND\n" +
                "M  V30 1 1 1 2\n" +
                "M  V30 2 3 2 3\n" +
                "M  V30 END BOND\n" +
                "M  V30 BEGIN SGROUP\n" +
                "M  V30 1 SUP 0 ATOMS=(2 2 3) SAP=(3 2 1 1) XBONDS=(1 1) LABEL=CN\n" +
                "M  V30 END SGROUP\n" +
                "M  V30 END CTAB\n" +
                "M  END\n";
        final Reader stringReader = new StringReader(v3000Mol);
        final MDLV3000Reader reader = new MDLV3000Reader(stringReader);
        final IAtomContainer atomContainer = reader.read(SilentChemObjectBuilder.getInstance().newAtomContainer());
        final ElectronDonation model = ElectronDonation.daylight();
        final CycleFinder cycles = Cycles.or(Cycles.all(), Cycles.all(6));
        final Aromaticity aromaticity = new Aromaticity(model, cycles);
        aromaticity.apply(atomContainer); // No NPE
        
        Aromaticity.cdkLegacy().apply(atomContainer); // throws NPE
        
    }
```

**Comment by johnmay** [comment]
CDK Aromaticity model depends on the atom type, pseudo atoms don't have atom types. The daylight model is generally preferred as it's more efficient.

**Comment by johnmay** [comment]
The actual bug here is the V3000 file is reading the SGROUP wrong. Abbreviations are actually just display short cuts: 

http://efficientbits.blogspot.com/2015/09/bringing-molfile-sgroups-to-cdk-demo.html

For example I don't think V2000 has the same issue?

```

  Mrv1810 07082011172D          

  3  2  0  0  0  0            999 V2000
   -2.7790    3.6998    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.0645    4.1123    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.3501    4.5248    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  3  0  0  0  0
M  STY  1   1 SUP
M  SAL   1  2   2   3
M  SBL   1  1   1
M  SMT   1 CN
M  SAP   1  1   2   1  1
M  END
```

**Comment by marois** [comment]
@johnmay I believe you're right about the V3000 reading being wrong. After submitting this issue I tried writing the above atomContainer as a v2000 file and no superatom (SUP) information is written because the MDLV3000Reader does not create an SGroup property, and probably should not replace the atom with a pseudo atom when dealing with abbreviations.

Also agree with using the Daylight model. We are switching to it, since we do not need to be backwards compatible. I initially thought this would solve our issue, but since we need to also export molecules in v2000 format I end up with CN as an atom:

```

  Mrv1810 07082011172D          

  3  2  0  0  0  0            999 V2000
   -2.7790    3.6998    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.0645    4.1123    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.3501    4.5248    0.0000 CN   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  3  0  0  0  0
```

Fortunately, `readSGroup` is public, so I can extend MDLV3000Reader to disable reading the "SUP" type for now and revert my changes when v2.4 becomes available. I'm going to need to do this anyway to bring in the `Stereo.UP_OR_DOWN` change in v2.4 `readBondBlock` code.

**Comment by johnmay** [comment]
I'm confused... please don't generate a file like this:

```

  Mrv1810 07082011172D          

  3  2  0  0  0  0            999 V2000
   -2.7790    3.6998    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.0645    4.1123    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.3501    4.5248    0.0000 CN   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  3  0  0  0  0
```

why do you think that is a good idea?

**Comment by johnmay** [comment]
What exactly are you trying to do? http://xyproblem.info/


I'm lost as to who's generating what - Marvin vs CDK. CDK will write the Sgroups if you add them - if you just put a random Pseudo atom in there it won't expand it though. This is consistent between V3000 and V2000.

**Comment by marois** [comment]
Sorry for the confusion. That V2000 output is generated by CDK due to the issue with the V3000 reader. Imagine in the test case above the `atomContainer` produced by the MDLV3000Reader were immediately written to a file using MDLV2000Writer.

**Comment by marois** [comment]
Meant to add I don't think it's a good idea either. Was just illustrating one of the issues caused by how MDLV3000Reader currently handles abbreviations.

Our problem is we have a compound registration system that accepts SD files in V2000 or V3000 format. We do the various operations like checking the molecules are unique and valid (e.g. don't include query bonds) and save the atomContainer in V2000 format because we render the molecules in the browser using MarvinJS, which in our configuration only accepts v2000.

**Comment by johnmay** [comment]
So if I understand correctly you're converting that V3000 to V2000 and you get:
```
  Mrv1810 07082011172D          

  3  2  0  0  0  0            999 V2000
   -2.7790    3.6998    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.0645    4.1123    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.3501    4.5248    0.0000 CN  0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  3  0  0  0  0
M   END
```

**Comment by marois** [comment]
Yes exactly

**Comment by johnmay** [comment]
OK that's a bug, will fix that :-)

**Comment by johnmay** [comment]
The sgroup should be preserved, and will fix the aromaticity issue as well.

**Comment by johnmay** [comment]
Should be fix on master now, #657
